src/net/socket.rs

```diff
@@ -0,0 +1,222 @@
+use crate::config::NetworkConfig;
+use crate::error::{NetworkError, NetworkErrorKind, NetworkResult};
+use crate::net::link_conditioner::LinkConditioner;
+use crate::net::{connection::ActiveConnections, events::SocketEvent};
+use crate::packet::Packet;
+use log::error;
+use mio::{Evented, Events, Poll, PollOpt, Ready, Token};
+use std::io;
+use std::mem;
+use std::net::{SocketAddr, ToSocketAddrs};
+use std::sync::{mpsc, Arc};
+
+const SOCKET: Token = Token(0);
+
+/// A reliable UDP socket implementation with configurable reliability and ordering guarantees.
+pub struct LaminarSocket {
+    socket: mio::net::UdpSocket,
+    config: Arc<NetworkConfig>,
+    connections: ActiveConnections,
+    recv_buffer: Vec<u8>,
+    link_conditioner: Option<LinkConditioner>,
+    event_sender: mpsc::Sender<SocketEvent>,
+    packet_receiver: mpsc::Receiver<Packet>,
+}
+
+impl LaminarSocket {
+    /// Binds to the socket and then sets up `ActiveConnections` to manage the "connections".
+    /// Because UDP connections are not persistent, we can only infer the status of the remote
+    /// endpoint by looking to see if they are still sending packets or not
+    pub fn bind<A: ToSocketAddrs>(
+        addresses: A,
```

Comment:
Hrm. I'm curious as to how it's a blocking call. Or, more to the point, can it be particularly slow?

Reply:
Since, `ToSocketAddr::to_socket_addrs` which extracts the socket can block, most APIs at this level of the transport are not concerned with dns res so we should be explicit this is not something we provide. It blocks because it makes a syscall to just get the socketaddr. This is why mio doesn't accept it.